[BC-breaking] C++ API parity: Linear

pbelevich opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #27309 C++ API parity: MultiheadAttention
* **#27382 C++ API parity: Linear**

-----

This PR is BC-breaking in the following way:

The arguments in `LinearOptions` are renamed to match the Python API:
- `in` -> `in_features`
- `out` -> `out_features`
- `with_bias` -> `bias`

The accessor calls (e.g. `in()`) in the user's code needs to be changed accordingly.

-----

Differential Revision: [D17766735](https://our.internmc.facebook.com/intern/diff/D17766735)

pbelevich committed:
C++ API parity: Linear

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Linear"


Differential Revision: [D17766735](https://our.internmc.facebook.com/intern/diff/D17766735)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Linear"


Differential Revision: [D17766735](https://our.internmc.facebook.com/intern/diff/D17766735)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Linear"


Differential Revision: [D17766735](https://our.internmc.facebook.com/intern/diff/D17766735)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Linear"


Differential Revision: [D17766735](https://our.internmc.facebook.com/intern/diff/D17766735)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Linear"


Differential Revision: [D17766735](https://our.internmc.facebook.com/intern/diff/D17766735)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Linear"


Differential Revision: [D17766735](https://our.internmc.facebook.com/intern/diff/D17766735)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Linear"


Differential Revision: [D17766735](https://our.internmc.facebook.com/intern/diff/D17766735)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Linear"


Differential Revision: [D17766735](https://our.internmc.facebook.com/intern/diff/D17766735)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Linear"


Differential Revision: [D17766735](https://our.internmc.facebook.com/intern/diff/D17766735)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Linear"


Differential Revision: [D17766735](https://our.internmc.facebook.com/intern/diff/D17766735)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Linear"


Differential Revision: [D17766735](https://our.internmc.facebook.com/intern/diff/D17766735)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Linear"


Differential Revision: [D17766735](https://our.internmc.facebook.com/intern/diff/D17766735)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Linear"


Differential Revision: [D17766735](https://our.internmc.facebook.com/intern/diff/D17766735)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Linear"


Differential Revision: [D17766735](https://our.internmc.facebook.com/intern/diff/D17766735)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Linear"


Differential Revision: [D17766735](https://our.internmc.facebook.com/intern/diff/D17766735)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Linear"


Differential Revision: [D17766735](https://our.internmc.facebook.com/intern/diff/D17766735)

[ghstack-poisoned]

yf225 commented on torch/csrc/api/include/torch/nn/functional.h:
```diff
@@ -4,3 +4,4 @@
 #include <torch/nn/functional/loss.h>
 #include <torch/nn/functional/pooling.h>
 #include <torch/nn/functional/embedding.h>
+#include <torch/nn/functional/linear.h>
```
nit: need new line here.

yf225 replied:
We should probably sort the file names alphabetically as well.

yf225 commented on torch/csrc/api/include/torch/nn/functional/linear.h:
```diff
@@ -0,0 +1,17 @@
+
+inline Tensor linear(const Tensor& input, const Tensor& weight,
+                     const Tensor& bias) {
+  return torch::linear(input, weight, bias);
```
I think we should strictly follow the Python implementation here:
```python
    if input.dim() == 2 and bias is not None:
        # fused op is marginally faster
        ret = torch.addmm(bias, input, weight.t())
    else:
        output = input.matmul(weight.t())
        if bias is not None:
            output += bias
        ret = output
    return ret
```
otherwise people would complain about difference in performance between Python and C++ code :(

yf225 commented on torch/csrc/api/include/torch/nn/functional/linear.h:
```diff
@@ -0,0 +1,17 @@
+namespace functional {
+
+inline Tensor linear(const Tensor& input, const Tensor& weight,
+                     const Tensor& bias) {
```
`bias` can be `None` in Python, and the equivalent default is `Tensor()` (aka. undefined tensor) in C++.

yf225 commented on torch/csrc/api/src/nn/modules/linear.cpp:
```diff
@@ -47,7 +50,7 @@ void LinearImpl::pretty_print(std::ostream& stream) const {
 
 Tensor LinearImpl::forward(const Tensor& input) {
   AT_ASSERT(!options.with_bias() || bias.defined());
```
We should remove this check to better match the Python version. :)

yf225 replied:
We would need some other changes to `Linear` as well:
1. Change `LinearOptions` to take `in_features` / `out_features` / `bias` instead of `in` / `out` / `with_bias`, to better match the Python version. (We would need to fix pretty_print as well.)
2. When `options.bias()` is false, we should still register the `bias` parameter, to better match the Python version.
3. The rest of the `reset` code should strictly follow:
```python
        init.kaiming_uniform_(self.weight, a=math.sqrt(5))
        if self.bias is not None:
            fan_in, _ = init._calculate_fan_in_and_fan_out(self.weight)
            bound = 1 / math.sqrt(fan_in)
            init.uniform_(self.bias, -bound, bound)
```

pbelevich committed:
Update on "C++ API parity: Linear"


Differential Revision: [D17766735](https://our.internmc.facebook.com/intern/diff/D17766735)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Linear"


Differential Revision: [D17766735](https://our.internmc.facebook.com/intern/diff/D17766735)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Linear"


Differential Revision: [D17766735](https://our.internmc.facebook.com/intern/diff/D17766735)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Linear"


Differential Revision: [D17766735](https://our.internmc.facebook.com/intern/diff/D17766735)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Linear"


Differential Revision: [D17766735](https://our.internmc.facebook.com/intern/diff/D17766735)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Linear"


Differential Revision: [D17766735](https://our.internmc.facebook.com/intern/diff/D17766735)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Linear"


Differential Revision: [D17766735](https://our.internmc.facebook.com/intern/diff/D17766735)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Linear"


Differential Revision: [D17766735](https://our.internmc.facebook.com/intern/diff/D17766735)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Linear"


Differential Revision: [D17766735](https://our.internmc.facebook.com/intern/diff/D17766735)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Linear"


Differential Revision: [D17766735](https://our.internmc.facebook.com/intern/diff/D17766735)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Linear"


Differential Revision: [D17766735](https://our.internmc.facebook.com/intern/diff/D17766735)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Linear"


Differential Revision: [D17766735](https://our.internmc.facebook.com/intern/diff/D17766735)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Linear"


Differential Revision: [D17766735](https://our.internmc.facebook.com/intern/diff/D17766735)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Linear"


Differential Revision: [D17766735](https://our.internmc.facebook.com/intern/diff/D17766735)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Linear"


Differential Revision: [D17766735](https://our.internmc.facebook.com/intern/diff/D17766735)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Linear"


Differential Revision: [D17766735](https://our.internmc.facebook.com/intern/diff/D17766735)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Linear"


Differential Revision: [D17766735](https://our.internmc.facebook.com/intern/diff/D17766735)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Linear"


Differential Revision: [D17766735](https://our.internmc.facebook.com/intern/diff/D17766735)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Linear"


Differential Revision: [D17766735](https://our.internmc.facebook.com/intern/diff/D17766735)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Linear"


Differential Revision: [D17766735](https://our.internmc.facebook.com/intern/diff/D17766735)

[ghstack-poisoned]

yf225 commented on torch/csrc/api/src/nn/init.cpp:
```diff
@@ -236,6 +236,29 @@ Tensor zeros_(Tensor tensor) {
+    "for tensor with fewer than 2 dimensions")
+  
+  int64_t fan_in, fan_out;
+  if (dimensions == 2) {
```
It would be awesome to add the comment in Python version: `  // Linear` here :D

yf225 commented on torch/csrc/api/include/torch/nn/functional/linear.h:
```diff
@@ -0,0 +1,26 @@
+#pragma once
+
+#include <torch/nn/options/activation.h>
```
Is there a reason we need to include `torch/nn/options/activation.h`?

yf225 commented on torch/csrc/api/include/torch/nn/functional.h:
```diff
@@ -1,7 +1,9 @@
 #pragma once
 
+#include <torch/nn/functional/activation.h>
```
Is there a reason we need to include `torch/nn/functional/activation.h`?

yf225 commented on torch/csrc/api/include/torch/nn/functional/linear.h:
```diff
@@ -0,0 +1,26 @@
+
+inline Tensor linear(const Tensor& input, const Tensor& weight,
+                     const Tensor& bias = {}) {
+  if (input.dim() == 2 and bias.defined()) {
```
We probably need to change it to
```suggestion
  if (input.dim() == 2 && bias.defined()) {
```

yf225 reviewed (commented):
Thanks a lot @pbelevich! I left some minor comments.

yf225 commented on torch/csrc/api/src/nn/modules/linear.cpp:
```diff
@@ -26,28 +30,33 @@ LinearImpl::LinearImpl(const LinearOptions& options_) : options(options_) {
+  if (options.bias()) {
+    bias = register_parameter("bias", torch::empty(options.out_features()));
+  } else {
+    bias = register_parameter("bias", {});
```
it would be awesome to also pass `/*requires_grad=*/false` into the `register_parameter()` function call, so that this doesn't print warning.

pbelevich replied:
Since [set_requires_grad is not implemented for Tensor](https://github.com/pytorch/pytorch/blob/master/c10/core/TensorImpl.h#L529) I cannot pass {} as argument to register_parameter() and I need to add another method like unsafe_ register_parameter which does not perform any checks and does not set_requires_grad

pbelevich committed:
Update on "C++ API parity: Linear"


Differential Revision: [D17766735](https://our.internmc.facebook.com/intern/diff/D17766735)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Linear"


Differential Revision: [D17766735](https://our.internmc.facebook.com/intern/diff/D17766735)

[ghstack-poisoned]

yf225 commented on torch/csrc/api/src/nn/modules/linear.cpp:
```diff
@@ -30,28 +30,33 @@ LinearImpl::LinearImpl(const LinearOptions& options_) : options(options_) {
-  NoGradGuard no_grad;
-  for (auto& p : this->parameters()) {
-    p.uniform_(-stdv, stdv);
+  torch::nn::init::kaiming_uniform_(weight, sqrt(5));
```
clang-tidy is acting up again :(
```
{
    path: 'torch/csrc/api/src/nn/modules/linear.cpp',
    start_line: 41,
    end_line: 41,
    start_column: 50,
    end_column: 50,
    annotation_level: 'failure',
    message: '[[cppcoreguidelines-avoid-magic-numbers]] warning: 5 is a ' +
      'magic number; consider replacing it with a named constant'
  }
```
Let's probably add `// NOLINT(cppcoreguidelines-avoid-magic-numbers)` at the end of this line :D

yf225 reviewed (approved):
@pbelevich Thanks a lot for the awesome work! We can land it after the fixing the annoying clang-tidy issue :D

pbelevich committed:
Update on "C++ API parity: Linear"


Differential Revision: [D17766735](https://our.internmc.facebook.com/intern/diff/D17766735)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Linear"


Differential Revision: [D17766735](https://our.internmc.facebook.com/intern/diff/D17766735)

[ghstack-poisoned]
